### PR TITLE
sstables/mx/writer: handle non-full prefix row keys

### PR DESCRIPTION
**Pull request by denesb**
Although valid for compact tables, non-full (or empty) clustering key prefixes are not handled for row keys when writing sstables. Only the present components are written, consequently if the key is empty, it is omitted entirely.
When parsing sstables, the parsing code unconditionally parses a full prefix.
This mis-match results in parsing failures, as the parser parses part of the row content as a key resulting in a garbage key and subsequent mis-parsing of the row content and maybe even subsequent partitions.

Introduce a new system table: `system.corrupt_data` and infrastructure similar to `large_data_handler`: `corrupt_data_handler` which abstracts how corrupt data is handled. The sstable writer now passes rows such corrupt keys to the corrupt data handler. This way, we avoid corrupting the sstables beyond parsing and the rows are also kept around in system.corrupt_data for later inspection and possible recovery.

Add a full-stack test which checks that rows with bad keys are correctly handled.

Fixes: https://github.com/scylladb/scylladb/issues/24489

The bug is present in all versions, has to be backported to all supported versions.

### CODE REVIEW

File: sstables/mx/writer.cc [hunk 1]
```diff
@@ -1291,7 +1291,19 @@ void writer::write_clustered(const clustering_row& clustered_row, uint64_t prev_
...
-    write_clustering_prefix(_sst.get_version(), *_data_writer, _schema, clustered_row.key(), ephemerally_full_prefix{_schema.is_compact_table()});
+    ephemerally_full_prefix is_ephemerally_full{_schema.is_compact_table()};
+    const auto& ckey = clustered_row.key();
+    if (!is_ephemerally_full && !ckey.is_full(_schema)) {
+        slogger.warn("writing non-full clustering key {} to partition {} in sstable {} of non-compact table {}.{}",
+                ckey,
+                _partition_key->to_partition_key(_schema),
+                _sst.get_filename(),
+                _schema.ks_name(),
+                _schema.cf_name());
```
[the remark this answers is not in the document]
Comment (denesb) [reply sstables/mx/writer.cc]: We don't expect a lot of null keys, so far it has been a very rare occurance so we want to see a warning for all.

**Comment by denesb** [comment]
The comment suppressed due to low confidence is actually more useful than the other one which was not suppressed: "Consider adding tests for non-full but non-empty clustering key prefixes (e.g., partial prefixes) to ensure the writer code path for those cases is also exercised."

**Comment by denesb** [comment]
New in v2: extend test to cover non-empty, non-full prefixes too.

File: test/boost/sstable_datafile_test.cc [hunk 2]
```diff
@@ -3278,3 +3278,63 @@ SEASTAR_TEST_CASE(sstable_identifier_correctness) {
...
 }
+
+SEASTAR_TEST_CASE(test_non_full_and_empty_row_keys) {
+    return test_env::do_with_async([] (test_env& env) {
+        auto random_spec = tests::make_random_schema_specification(
+                get_name(),
+                std::uniform_int_distribution<size_t>(1, 4),
+                std::uniform_int_distribution<size_t>(1, 4),
+                std::uniform_int_distribution<size_t>(2, 8),
+                std::uniform_int_distribution<size_t>(2, 8));
```
Comment (avikivity) [inline test/boost/sstable_datafile_test.cc:3293]: random_note: designated initializers would be nicer here

File: test/boost/sstable_datafile_test.cc [hunk 3]
```diff
@@ -3278,3 +3278,63 @@ SEASTAR_TEST_CASE(sstable_identifier_correctness) {
...
+                return max_ts;
+            }
+        });
+
+        const auto source_mutation = mutation_description.build(schema);
+
+        auto add_row_with_patched_key = [&] (const clustering_key& ckey) {
+            testlog.info("Adding row with patched key: {}", ckey);
+            clustering_row row(ckey, deletable_row(*schema, source_mutation.partition().clustered_rows().begin()->row()));
+            mutations.back().apply(mutation_fragment(*schema, env.make_reader_permit(), std::move(row)));
```
Comment (avikivity) [inline test/boost/sstable_datafile_test.cc]: Is there no way to generate these with CQL?

Comment (denesb) [reply test/boost/sstable_datafile_test.cc]: I tried but did not find a way. If there was we could just plug that hole and makr this patch throw.

Comment (avikivity) [reply test/boost/sstable_datafile_test.cc]: What I meant here is that if there's no way to generate those keys, we shouldn't handle them. It's better to drop the row with the bad key than change the key to something else.

Comment (denesb) [reply test/boost/sstable_datafile_test.cc]: I considered this and decided to write the row, so the content can be recovered later if the case may be. We can make scrub discard such rows. I agree the use-case is marginal, in many cases the content is useless without the key, so I don't insist.

Comment (denesb) [reply test/boost/sstable_datafile_test.cc]: Another point to the above: rows with non-full keys cannot be point-queried via CQL, but they can be retrieved by partition scans.

File: test/boost/sstable_datafile_test.cc (hunk 2 above)
Comment (denesb) [reply test/boost/sstable_datafile_test.cc:3293]: Yes

**Comment by denesb** [comment]
New in v3:
* Don't generate range tombstones in the test -- they cause complications because bad rows violate assumptions about row keys being full and can reorder w.r.t. range tombstone change fragments.

**Comment by denesb** [comment]
New in v3.1: fix copy-pasta copying a return statement together with comment.

File: test/boost/sstable_datafile_test.cc (hunk 3 above)
Comment (avikivity) [reply test/boost/sstable_datafile_test.cc]: Maybe we should write it somewhere else (like scrub recovers rows). But we can't write a row that the user hasn't written into the database. It will get propagated by read repair and infect other replicas.

Comment (denesb) [reply test/boost/sstable_datafile_test.cc]: I spent some more time thinking about this, and also some time figthing the failing unit test, because range tombstones don't work well with rows with empty keys.
I now think that a warning and writing the row into the SSTable is too tame of a response. We are dealing with corruption here that requires immediate response, it should trigger alarms. I will change the warning to error.
I also agree that writing this to the SSTable is dangerous, maybe now is a good time to revive the idea of the `system.lost_and_found` table.

Comment (denesb) [reply test/boost/sstable_datafile_test.cc]: One more thing: I think we should have an `seastar_internal_errors` metric, which is bumped every time we hit `on_internal_error()`. Then we should set up alerts for any time this is incremented.
Hitting an `on_internal_error()` is always a bad thing and somebody should have a look.

**Comment by denesb** [comment]
> [test_random_failures[stop_after_starting_migration_manager-add_new_udt].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17724/testReport/junit/cluster.random_failures/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_after_starting_migration_manager_add_new_udt__debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures%5Bstop_after_starting_migration_manager-add_new_udt%5D.debug.2)

https://github.com/scylladb/scylladb/issues/24513

File: test/boost/sstable_datafile_test.cc (hunk 3 above)
Comment (denesb) [reply test/boost/sstable_datafile_test.cc]: https://github.com/scylladb/seastar/pull/2809

Comment (bhalevy) [reply test/boost/sstable_datafile_test.cc]: We may also want to change the internal error format to indicate that it is an internal error.
I don't think we explicitly say that on the log message.

**Comment by avikivity** [comment]
> > [test_random_failures[stop_after_starting_migration_manager-add_new_udt].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17724/testReport/junit/cluster.random_failures/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_after_starting_migration_manager_add_new_udt__debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures%5Bstop_after_starting_migration_manager-add_new_udt%5D.debug.2)
> 
> #24513

Rebase will fix, offending patch was reverted

**Comment by denesb** [comment]
> > > [test_random_failures[stop_after_starting_migration_manager-add_new_udt].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17724/testReport/junit/cluster.random_failures/test_random_failures/Tests___Unit_Tests___test_random_failures_stop_after_starting_migration_manager_add_new_udt__debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures%5Bstop_after_starting_migration_manager-add_new_udt%5D.debug.2)
> > 
> > 
> > #24513
> 
> Rebase will fix, offending patch was reverted

I want to send a new version, will also do a rebase.

**Comment by denesb** [comment]
New in v4:
* Promote the warning to error.
* Introduce `system.corrupt_data` and `corrupt_data_handler` to handle corrupt rows.

File: test/boost/sstable_datafile_test.cc (hunk 3 above)
Comment (denesb) [reply test/boost/sstable_datafile_test.cc]: Fixed in v4.

**Comment by denesb** [comment]
> #### 🟢 CI State: SUCCESS
> #### Build Details:
> 
>     * Duration: 4 min 2 sec
> 
>     * Builder: spider6.cloudius-systems.com

@yaronkaikov I don't get this, successful CI in 4 minutes? Also, no details on individual tests.

**Comment by xtrey** [comment]
> > #### 🟢 CI State: SUCCESS
> > #### Build Details:
> > ```
> > * Duration: 4 min 2 sec
> > 
> > * Builder: spider6.cloudius-systems.com
> > ```
> 
> @yaronkaikov I don't get this, successful CI in 4 minutes? Also, no details on individual tests.

It failed on checkout step before any of test steps was able to start, hence no information. Seems some issue with node, because there are several more failed jobs with the same error on checkout step.

**Comment by yaronkaikov** [comment]
@denesb the issue was indeed with `spider6` , at the mean time, i re-kicked the CI - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17814/

File: idl/mutation.idl.hh [hunk 4]
```diff
@@ -168,3 +175,7 @@ class mutation_fragment stub [[writable]] {
                    partition_start, partition_end> fragment;
 };
 
+class mutation_fragment_v2 stub [[writable]] {
+    std::variant<clustering_row, static_row, range_tombstone_change,
+                   partition_start, partition_end> fragment;
+};
```
Comment (avikivity) [inline idl/mutation.idl.hh:181]: We should start using it in rpc.

Comment (avikivity) [reply idl/mutation.idl.hh:181]: But didn't we have something similar when we changed the half-reversed reads?

Comment (avikivity) [reply idl/mutation.idl.hh:181]: ref: 8c257db28322a134e131ae2200200c6f878a5c8a

Comment (avikivity) [reply idl/mutation.idl.hh:181]: I guess that was still using the old frozen mutation fragment. @paszkow

**Comment by avikivity** [comment]
Looks good. Need test so the code that records bad records isn't exercised for the first time in production. Need a metrics so we can set an alert for it.

@amnonh let's have a file that lists all metrics that are required to be zero and when they aren't we want an alert.

**Comment by amnonh** [comment]
> Looks good. Need test so the code that records bad records isn't exercised for the first time in production. Need a metrics so we can set an alert for it.
> 
> @amnonh let's have a file that lists all metrics that are required to be zero and when they aren't we want an alert.

Only if we use it to auto-generate alerts and documentation, if not, it's better to open an issue in the monitoring repository, and I'll manually add one

**Comment by avikivity** [comment]
> > Looks good. Need test so the code that records bad records isn't exercised for the first time in production. Need a metrics so we can set an alert for it.
> > @amnonh let's have a file that lists all metrics that are required to be zero and when they aren't we want an alert.
> 
> Only if we use it to auto-generate alerts and documentation, if not, it's better to open an issue in the monitoring repository, and I'll manually add one

Let's use it to auto-generate alerts.

**Comment by avikivity** [comment]
Or perhaps, we can attach a label to such metrics. Though I'm not fond of label overflow.

**Comment by amnonh** [comment]
> Or perhaps, we can attach a label to such metrics. Though I'm not fond of label overflow.

That may cause an issue with counters where non-zero is the metric derivative

**Comment by denesb** [comment]
> Looks good. Need test so the code that records bad records isn't exercised for the first time in production.

I included a full-stack test, which generates random mutations, and adds one bad rows, then checks that the sstable only contains the non-bad row and that the `system.corrupt_data`  contains the expected entry and all the fields have the correct values.
The test is included in the last commit (and is mentioned in the cover-letter too).

File: idl/mutation.idl.hh (hunk 4 above)
Comment (denesb) [reply idl/mutation.idl.hh:181]: This has no connection with reversed reads, those use the query-result to transmit results to the coordinator.
The only user of `frozen_mutation_fragment` (v1) is repair and streaming and indeed it is high time to switch to v2, so we can start the deprecation clock for v1 mutation fragment, which is left alive just for these two users.

**Comment by denesb** [comment]
New in v5:
* Split test into a commit of its own (so it is more visible)
* Extend test to check stats too
* Add metric for registered corrupt data

**Comment by denesb** [comment]
> Looks good. Need test so the code that records bad records isn't exercised for the first time in production. Need a metrics so we can set an alert for it.

Added metric in v5.

**Comment by denesb** [comment]
@avikivity ping for re-review/merge.

File: idl/mutation.idl.hh (hunk 4 above)
Comment (avikivity) [reply idl/mutation.idl.hh:181]: Right. @asias
